@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -736,105 +747,79 @@ export class TiltViewClipPlanesEvaluator extends TopViewClipPlanesEvaluator {
-            // where:
-            // |t| = sqrt(|d|^2 - |r|^2), or
-            // |t| = cos(alpha) * |d|
-            // By simplifying t_d equation with the second condition, we get:
-            // t_d = d_norm * cos(alpha) * |d|, where d = d_norm * |d|
-            // t_d = d * cos(alpha)
-            // Cause cameraToOrigin is no longer needed re-use it to calculate td.
-            td = cameraToOrigin.multiplyScalar(cosAlpha);
-        }
-        // Second case takes into account the elevation above the ground.
-        else {
-            // Here the length of the tangent is extended with 'te' vector that allows to see
-            // elevated geometry beyond the tangent (horizon line), see
-            // TopViewClipPlanesEvaluator for explanations.
-            // t_d = d_norm * |t + te|,
-            // where:
-            // |t| = cos(alpha) * |d|
-            // |te| = sqrt((r+e)^2 - r^2)
-            // Which reduces to:
-            // |te| = sqrt(2*r*e + e^2)
-            const t = cosAlpha * d;
-            const te = Math.sqrt(2 * r * this.maxElevation - this.maxElevation * this.maxElevation);
-            // Re-use pre-allocated vector.
-            td = cameraToOrigin.copy(dNorm).multiplyScalar(t + te);
-        }
-        // For tangent, oriented in the direction to origin, we then apply
-        // rotation to it on every rotation axes using already known tangent angle,
-        // the angle that defines maximum visibility along the sphere surface.
-        // This gives us the tangent rays in all major directions.
-        const ry = this.m_tmpQuaternion.setFromAxisAngle(yaxis.cross(dNorm), alpha);
-        const tdry = td.clone().applyQuaternion(ry);
-
-        const rx = this.m_tmpQuaternion.setFromAxisAngle(xaxis.cross(dNorm), alpha);
-        const tdrx = td.clone().applyQuaternion(rx);
-
-        const rx2 = this.m_tmpQuaternion.copy(rx).inverse();
-        // Cause td vector in no longer needed, reuse it applying quaternion to it.
-        const tdrx2 = td.applyQuaternion(rx2);
-
-        // Rotated tangents are then added to camera position thus defining far plane
-        // position and orientation - it is enough to have three co-planar points to
-        // define the plane.
-        // p1 = camera.position + tdrx
-        // p2 = camera.position + tdry
-        // p3 = camera.position + tdrx2
-        // const farPlane = new THREE.Plane().setFromCoplanarPoints(p1, p2, p3);
-        // viewRanges.far = farPlane.distanceToPoint(camera.position);
-
-        // This of course may be simplified by moving calculus entirely to camera space,
-        // because far plane is indeed defined in that space anyway:
-        const farPlane = this.m_tmpPlane.setFromCoplanarPoints(tdrx, tdry, tdrx2);
-        viewRanges.far = farPlane.constant;
-        // Take into account largest depression assumed, that may be further then
-        // tangent based far plane distance.
+        const sinAlpha2 = r / (r + cameraAltitude);
```

Comment:
Could check how much '(r + cameraAltitude)' differs from 'cameraToOrigin.length()'  (before normalization), because it should be exactly equal... Maybe that is the reason of problems...

Reply:
Yes this can be simplified to cameraToOrigin.length().